diff --git a/include/driver/tool.h b/include/driver/tool.h
index 93b7bb61..4c3ff080 100644
--- a/include/driver/tool.h
+++ b/include/driver/tool.h
@@ -70,6 +70,9 @@ struct DriverToolOptions {
             "Enable generating code for all statistics options include instruction counting, gas measuring, and execution time"sv)),
         ConfForceInterpreter(
             PO::Description("Forcibly run WASM in interpreter mode."sv)),
+        DumpFlag(PO::Description("Unsnapshot statement to img files."sv)),
+        RestoreFlag(PO::Description("Restore statement by img files."sv)),
+        ImageDir(PO::Description("Restore statement by img files."sv), PO::MetaVar("IMAGE_DIR"sv)),
         TimeLim(
             PO::Description(
                 "Limitation of maximum time(in milliseconds) for execution, default value is 0 for no limitations"sv),
@@ -107,6 +110,10 @@ struct DriverToolOptions {
   PO::Option<PO::Toggle> ConfEnableTimeMeasuring;
   PO::Option<PO::Toggle> ConfEnableAllStatistics;
   PO::Option<PO::Toggle> ConfForceInterpreter;
+  PO::Option<PO::Toggle> DumpFlag;
+  PO::Option<PO::Toggle> RestoreFlag;
+  PO::List<std::string> ImageDir;
+  PO::Option<PO::Toggle> DebugMode;
   PO::Option<uint64_t> TimeLim;
   PO::List<int> GasLim;
   PO::List<int> MemLim;
@@ -136,6 +143,10 @@ struct DriverToolOptions {
         .add_option("enable-extended-const"sv, PropExtendConst)
         .add_option("enable-threads"sv, PropThreads)
         .add_option("enable-all"sv, PropAll)
+        .add_option("no-snapshot"sv, DumpFlag)
+        .add_option("restore"sv, RestoreFlag)
+        .add_option("image-dir"sv, ImageDir)
+        .add_option("debug-mode"sv, DebugMode)
         .add_option("time-limit"sv, TimeLim)
         .add_option("gas-limit"sv, GasLim)
         .add_option("memory-page-limit"sv, MemLim)
