diff --git a/include/common/statistics.h b/include/common/statistics.h
index d06f8460..175e35f7 100644
--- a/include/common/statistics.h
+++ b/include/common/statistics.h
@@ -182,12 +182,27 @@ public:
     }
   }
 
+  // DumpFlag
+  void setDumpFlag(bool f) { DumpFlag = f; }
+  bool getDumpFlag()       { return DumpFlag; }
+  
+  // RestoreFlag
+  void setRestoreFlag(bool f) { RestoreFlag = f; }
+  bool getRestoreFlag()       { return RestoreFlag; }
+  
+  // DebugMode
+  void setDebugMode(bool f) { DebugMode = f; }
+  bool getDebugMode()       { return DebugMode; }
+
 private:
   std::vector<uint64_t> CostTab;
   std::atomic_uint64_t InstrCnt;
   uint64_t CostLimit;
   std::atomic_uint64_t CostSum;
   Timer::Timer TimeRecorder;
+  bool DumpFlag;
+  bool RestoreFlag;
+  bool DebugMode;
 };
 
 } // namespace Statistics
